sql: Ensure we auto commit for first DDL in txn

spilchen opened the pull request:
Previously, when the autocommit_before_ddl session setting was enabled, the first DDL statement in a multi-statement transaction did not trigger an auto-commit, leading to the use of the legacy schema changer. This behavior was observed in transactions such as:

```
SET autocommit_before_ddl = ON;
BEGIN;
DROP TABLE t1;
DROP TABLE t2;
```

This change will now ensure that we auto commit before the first drop. This ensures that we'll use the DSC for both drops.

Epic: None
Release note: None

Closes #141034

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141145)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @asg0451 and @jeffswenson)

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTr3mX04iMI-pTLv4Y:-OJTr3mX04iMI-pTLv4Z:bun9uly) (no related file):*
the logic test changes lgtm overall minus some small nits, but i'm just curious if you had to track those down manually, or if the code change made the tests start failing?
___
*[`pkg/sql/conn_executor_ddl.go` line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJThJmC9XzmKLcraw2r:-OJThJmC9XzmKLcraw2s:bynzkhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/conn_executor_ddl.go#L39)):*
> ```Go
> 	ctx context.Context, ast tree.Statement,
> ) (fsm.Event, fsm.EventPayload) {
> 	// We will auto-commit in the following cases:
> ```

super nit: could we add something in the comment to make it clear these conditions must _all_ be true. (not that at least one of the conditions is true.)
___
*[`pkg/sql/conn_executor_ddl.go` line 50 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTpW13AoJrXhcajkBP:-OJTpW13AoJrXhcajkBQ:b-81o9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/conn_executor_ddl.go#L50)):*
> ```Go
> 		ex.sessionData().AutoCommitBeforeDDL &&
> 		(!ex.planner.EvalContext().TxnIsSingleStmt || !ex.implicitTxn()) &&
> 		(ex.extraTxnState.firstStmtExecuted || !ex.implicitTxn()) {
> ```

nit: do you think this would be more clear if we wrote it as:
```
	if ex.executorType != executorTypeInternal &&
		tree.CanModifySchema(ast) &&
		ex.sessionData().AutoCommitBeforeDDL &&
		(!ex.implicitTxn() || (!ex.planner.EvalContext().TxnIsSingleStmt && ex.extraTxnState.firstStmtExecuted)) {
```

and to enhance readability further, we could define local variables too:
```
	explicitTxn := !ex.implicitTxn()
	multiStmtTxn := !ex.planner.EvalContext().TxnIsSingleStmt
```

then we'd end up with
```
	if ex.executorType != executorTypeInternal &&
		tree.CanModifySchema(ast) &&
		ex.sessionData().AutoCommitBeforeDDL &&
		(explicitTxn || (multiStmtTxn && ex.extraTxnState.firstStmtExecuted)) {
```
___
*[`pkg/backup/tenant_backup_nemesis_test.go` line 155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTt6zDC7vWIx3O-jAn:-OJTt6zDC7vWIx3O-jAo:bevskes) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/backup/tenant_backup_nemesis_test.go#L155)):*
> ```Go
> 	require.NoError(t, err)
> 	tenant10Conn := tenant10.SQLConn(t, serverutils.DBName("defaultdb"))
> 	_, err = tenant10Conn.Exec("SET autocommit_before_ddl=off")
> ```

super nit: i'm curious why this change was needed
___
*[`pkg/sql/logictest/testdata/logic_test/strict_ddl_atomicity` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTrruP0S_CWSQWsTuz:-OJTrruP0S_CWSQWsTv-:bvswcd9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/strict_ddl_atomicity#L2)):*
> ```
> # This test requires autocommit_before_ddl to be off, but it cannot be off for repeatable read and read committed isolation levels.
> # LogicTest: !local-repeatable-read !local-read-committed
> ```

nit: let's use the `weak-iso-level-configs` alias. or similar to above, could we just specify `BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;` explicitly?
___
*[`pkg/sql/logictest/testdata/logic_test/strict_ddl_atomicity` line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTrwck3tMwdbPBhWX_:-OJTrwck3tMwdbPBhWXa:bgcns17) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/strict_ddl_atomicity#L15)):*
> ```
> 
> statement ok
> SET LOCAL autocommit_before_ddl=off;
> ```

if this should be off for the whole test, then we can set it once with `SET` (without `local`). using `SET LOCAL` means the change is only applied within the current open txn.
___
*[`pkg/sql/logictest/testdata/logic_test/discard` line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTqROq6yT9Zk8WRZt_:-OJTqROq6yT9Zk8WRZta:b-q7sd95) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/discard#L58)):*
> ```
> SET LOCAL autocommit_before_ddl=off;
> 
> skipif config local-read-committed local-repeatable-read
> ```

nit: there is a "config set" that can be used as an alias: `weak-iso-level-configs`

https://github.com/cockroachdb/cockroach/blob/8075340257159d8f634a3333c36e1e54fcaaedec/pkg/sql/logictest/logictestbase/logictestbase.go#L626


though in this case, i'd rather fix the test by changing the `BEGIN` to `BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;`. we already have other tests for auto-upgrading the isolation level for weaker isolation level configs.
___
*[`pkg/sql/logictest/testdata/logic_test/txn` line 1489 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTsIh82pa9EkPhzEK5:-OJTsIh82pa9EkPhzEK6:b-pt6vah) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/txn#L1489)):*
> ```
> COMMIT
> 
> # CREATE SEQUENCE below will autocommit, so ensure the default is writes
> ```

could you explain this change a bit more? shouldn't we keep the `BEGIN READ WRITE`?
___
*[`pkg/sql/logictest/testdata/logic_test/expression_index` line 820 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTqegz1pxmgFhmSlMK:-OJTqegz1pxmgFhmSlML:btb2abf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/expression_index#L820)):*
> ```
> SET LOCAL autocommit_before_ddl=off;
> 
> skipif config local-read-committed local-repeatable-read
> ```

instead of having different assertions under weaker isolation levels, i'd rather fix the test by changing the `BEGIN` to `BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;`
___
*[`pkg/sql/logictest/testdata/logic_test/expression_index` line 1230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTrHkE20fuW3sOY2M7:-OJTrHkE20fuW3sOY2M8:btb2abf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/expression_index#L1230)):*
> ```
> skipif config local-repeatable-read local-read-committed
> statement ok
> BEGIN;
> ```

instead of having different assertions under weaker isolation levels, i'd rather fix the test by changing the `BEGIN` to `BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;`
___
*[`pkg/sql/testdata/savepoints` line 345 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTtwT34IgqJRtTYbEJ:-OJTtwT34IgqJRtTYbEK:br8bwpm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/testdata/savepoints#L345)):*
> ```
> sql
> BEGIN; CREATE TABLE unused(x INT)
> SAVEPOINT foo
> ```

nit: maybe the test comment should be updated to mention that when autocommit_before_ddl is on, we ignore SAVEPOINT statements that appear outside of a transaction.

for your reference, that's here: https://github.com/cockroachdb/cockroach/blob/500b4fc8ecaf92b392fdc291fa5b40a6bf6fdde1/pkg/sql/conn_executor_exec.go#L3481-L3483
___
*[`pkg/sql/logictest/testdata/logic_test/schema_change_in_txn` line 2055 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTrWEpDPxD-xlMg9k3:-OJTrWEpDPxD-xlMg9k4:bl979w1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/schema_change_in_txn#L2055)):*
> ```
> 
> statement ok
> BEGIN;
> ```

instead of having different assertions under weaker isolation levels, i'd rather change the test from using `BEGIN` to `BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;`
___
*[`pkg/sql/logictest/testdata/logic_test/manual_retry` line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTrPKZCj2BAeBlYXxd:-OJTrPKZCj2BAeBlYXxe:btb2abf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/manual_retry#L125)):*
> ```
> SAVEPOINT cockroach_restart
> 
> skipif config local-repeatable-read local-read-committed
> ```

instead of having different assertions under weaker isolation levels, i'd rather fix the test by changing the `BEGIN` to `BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;`


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141145)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @asg0451, @jeffswenson, and @spilchen)

___
*[`pkg/sql/conn_executor_ddl.go` line 50 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTpW13AoJrXhcajkBP:-OJTwtBJ3Xz_jWNGEbIH:bpfki2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/conn_executor_ddl.go#L50)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: do you think this would be more clear if we wrote it as:
```
	if ex.executorType != executorTypeInternal &&
		tree.CanModifySchema(ast) &&
		ex.sessionData().AutoCommitBeforeDDL &&
		(!ex.implicitTxn() || (!ex.planner.EvalContext().TxnIsSingleStmt && ex.extraTxnState.firstStmtExecuted)) {
```

and to enhance readability further, we could define local variables too:
```
	explicitTxn := !ex.implicitTxn()
	multiStmtTxn := !ex.planner.EvalContext().TxnIsSingleStmt
```

then we'd end up with
```
	if ex.executorType != executorTypeInternal &&
		tree.CanModifySchema(ast) &&
		ex.sessionData().AutoCommitBeforeDDL &&
		(explicitTxn || (multiStmtTxn && ex.extraTxnState.firstStmtExecuted)) {
```
</blockquote></details>

actually, is the `multiStmtTxn` variable even necessary now? according to boolean logic, it is still needed, but based on the semantic meaning of these variables, could we just have:
```
	if ex.executorType != executorTypeInternal &&
		tree.CanModifySchema(ast) &&
		ex.sessionData().AutoCommitBeforeDDL &&
		(explicitTxn || ex.extraTxnState.firstStmtExecuted) {
```


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141145)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @asg0451, @jeffswenson, and @rafiss)

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTr3mX04iMI-pTLv4Y:-OJUIfgF3VKocezvLjdu:bit37sk) (no related file):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

the logic test changes lgtm overall minus some small nits, but i'm just curious if you had to track those down manually, or if the code change made the tests start failing?
</blockquote></details>

No, it wasn’t done manually. I relied on the tests failing to identify them.
___
*[`pkg/sql/conn_executor_ddl.go` line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJThJmC9XzmKLcraw2r:-OJYTu-Q8T2DyK0P-Au9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/conn_executor_ddl.go#L39)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: could we add something in the comment to make it clear these conditions must _all_ be true. (not that at least one of the conditions is true.)
</blockquote></details>

Done.
___
*[`pkg/sql/conn_executor_ddl.go` line 50 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTpW13AoJrXhcajkBP:-OJYTwEA01fIM6IH4yym:bgwouqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/conn_executor_ddl.go#L50)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

actually, is the `multiStmtTxn` variable even necessary now? according to boolean logic, it is still needed, but based on the semantic meaning of these variables, could we just have:
```
	if ex.executorType != executorTypeInternal &&
		tree.CanModifySchema(ast) &&
		ex.sessionData().AutoCommitBeforeDDL &&
		(explicitTxn || ex.extraTxnState.firstStmtExecuted) {
```
</blockquote></details>

I like your suggestions. And simplifying the check seemed to work fine in my local testing.
___
*[`pkg/sql/logictest/testdata/logic_test/expression_index` line 820 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTqegz1pxmgFhmSlMK:-OJUWoSxAvVeRRf-taYc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/expression_index#L820)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

instead of having different assertions under weaker isolation levels, i'd rather fix the test by changing the `BEGIN` to `BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;`
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/expression_index` line 1230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTrHkE20fuW3sOY2M7:-OJUWph641DqUlpyU9uR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/expression_index#L1230)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

instead of having different assertions under weaker isolation levels, i'd rather fix the test by changing the `BEGIN` to `BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;`
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/manual_retry` line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTrPKZCj2BAeBlYXxd:-OJUX3WyBadKBSnNAwpg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/manual_retry#L125)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

instead of having different assertions under weaker isolation levels, i'd rather fix the test by changing the `BEGIN` to `BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;`
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/schema_change_in_txn` line 2055 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTrWEpDPxD-xlMg9k3:-OJUek8U-GKCXGRk5OVg:bypbvuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/schema_change_in_txn#L2055)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

instead of having different assertions under weaker isolation levels, i'd rather change the test from using `BEGIN` to `BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;`
</blockquote></details>

If the isolation starts as serializable, the autocommit will start a new transaction at a weaker isolation level, causing the extra NOTICE be displayed. I'm not sure how to get around that. So, I opted to set the default transaction isolation to serializable for this subtest.
___
*[`pkg/sql/logictest/testdata/logic_test/strict_ddl_atomicity` line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTrwck3tMwdbPBhWX_:-OJUfw1R9ix-bGzgBZXJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/strict_ddl_atomicity#L15)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

if this should be off for the whole test, then we can set it once with `SET` (without `local`). using `SET LOCAL` means the change is only applied within the current open txn.
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/txn` line 1489 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTsIh82pa9EkPhzEK5:-OJUgBgGBxkWq-lhB97e:bsex28h) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/txn#L1489)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could you explain this change a bit more? shouldn't we keep the `BEGIN READ WRITE`?
</blockquote></details>

Maybe this requires another code change. Consider the following sequence of commands:

```
SET default_transaction_read_only = true; 
BEGIN READ WRITE; 
CREATE SEQUENCE s;
```

The `CREATE SEQUENCE` statement commits, but the new transaction starts as read-only due to `default_transaction_read_only`. Should we ensure that when auto-committing, the new transaction inherits the read/write state of the one just committed?
___
*[`pkg/sql/testdata/savepoints` line 345 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTtwT34IgqJRtTYbEJ:-OJUi-rT4tudLjqyqMo1:b-xeczvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/testdata/savepoints#L345)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: maybe the test comment should be updated to mention that when autocommit_before_ddl is on, we ignore SAVEPOINT statements that appear outside of a transaction.

for your reference, that's here: https://github.com/cockroachdb/cockroach/blob/500b4fc8ecaf92b392fdc291fa5b40a6bf6fdde1/pkg/sql/conn_executor_exec.go#L3481-L3483
</blockquote></details>

Good point. I'll update the test comment.
___
*[`pkg/backup/tenant_backup_nemesis_test.go` line 155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTt6zDC7vWIx3O-jAn:-OJUi3kC1a0yJUOturis:brk0wcr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/backup/tenant_backup_nemesis_test.go#L155)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: i'm curious why this change was needed
</blockquote></details>

This was a strange one. Without it the test failed in this manner:
```
        	Error Trace:	pkg/backup/tenant_backup_nemesis_test.go:171
        	Error:      	Received unexpected error:
        	            	failed insert into bank: pq: relation "bank" does not exist
```

The test will setup the connection in the following way:
```
	tenant10Conn := tenant10.SQLConn(t, serverutils.DBName("defaultdb"))
	_, err = tenant10Conn.Exec("CREATE DATABASE bank")
	require.NoError(t, err)
	_, err = tenant10Conn.Exec("USE bank")
	require.NoError(t, err)
```

The error occurs when data is ingested concurrently using goroutines. I suspect that since ingestion spawns new connections from the connection pool, they default to `defaultdb` and cannot find the `bank` table.

I'm still unsure how `autocommit_before_ddl` affects this, as I would expect the same behaviour with it enabled. To address the issue, I removed the autocommit setting and reinitialized `tenant10Conn`, specifying `bank` as the database name.

So, now it looks like this:
```
	tenant10Conn := tenant10.SQLConn(t, serverutils.DBName("defaultdb"))
	_, err = tenant10Conn.Exec("CREATE DATABASE bank")
	require.NoError(t, err)
	tenant10Conn = tenant10.SQLConn(t, serverutils.DBName("bank"))
```


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141145)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @asg0451, @jeffswenson, and @spilchen)

___
*[`pkg/sql/logictest/testdata/logic_test/txn` line 1489 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTsIh82pa9EkPhzEK5:-OJZzsPs-_YORteyrCoD:b-i0kcn6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/txn#L1489)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Maybe this requires another code change. Consider the following sequence of commands:

```
SET default_transaction_read_only = true; 
BEGIN READ WRITE; 
CREATE SEQUENCE s;
```

The `CREATE SEQUENCE` statement commits, but the new transaction starts as read-only due to `default_transaction_read_only`. Should we ensure that when auto-committing, the new transaction inherits the read/write state of the one just committed?
</blockquote></details>

ah i see, thanks for explaining. what if we define the test like this:
```
diff --git a/pkg/sql/logictest/testdata/logic_test/txn b/pkg/sql/logictest/testdata/logic_test/txn
index 772a0ad1c6f..b7aa814fe07 100644
--- a/pkg/sql/logictest/testdata/logic_test/txn
+++ b/pkg/sql/logictest/testdata/logic_test/txn
@@ -1488,7 +1488,8 @@ COMMIT

 # BEGIN READ WRITE overrides READ ONLY default
 statement ok
-BEGIN READ WRITE
+BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE READ WRITE;
+SET LOCAL autocommit_before_ddl = false;

 statement ok
 CREATE SEQUENCE a
```

the intention of this test was to ensure that `BEGIN ... READ WRITE;` overrides the `SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY` that was configured earlier in this test, so i think we should preserve that intention
___
*[`pkg/backup/tenant_backup_nemesis_test.go` line 155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTt6zDC7vWIx3O-jAn:-OJ_-DrS79Xuc0DXi88F:b-sob57q) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/backup/tenant_backup_nemesis_test.go#L155)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

This was a strange one. Without it the test failed in this manner:
```
        	Error Trace:	pkg/backup/tenant_backup_nemesis_test.go:171
        	Error:      	Received unexpected error:
        	            	failed insert into bank: pq: relation "bank" does not exist
```

The test will setup the connection in the following way:
```
	tenant10Conn := tenant10.SQLConn(t, serverutils.DBName("defaultdb"))
	_, err = tenant10Conn.Exec("CREATE DATABASE bank")
	require.NoError(t, err)
	_, err = tenant10Conn.Exec("USE bank")
	require.NoError(t, err)
```

The error occurs when data is ingested concurrently using goroutines. I suspect that since ingestion spawns new connections from the connection pool, they default to `defaultdb` and cannot find the `bank` table.

I'm still unsure how `autocommit_before_ddl` affects this, as I would expect the same behaviour with it enabled. To address the issue, I removed the autocommit setting and reinitialized `tenant10Conn`, specifying `bank` as the database name.

So, now it looks like this:
```
	tenant10Conn := tenant10.SQLConn(t, serverutils.DBName("defaultdb"))
	_, err = tenant10Conn.Exec("CREATE DATABASE bank")
	require.NoError(t, err)
	tenant10Conn = tenant10.SQLConn(t, serverutils.DBName("bank"))
```
</blockquote></details>

thanks! i think your latest change captures the intent of the test better
___
*[`pkg/sql/logictest/testdata/logic_test/strict_ddl_atomicity` line 13 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJZziOX9QY5oeTpG7u1:-OJZziOYAiTQ6qRXv0fl:b-v3xigy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef032150ea632680d52f7e29f8fe1fe4d5dbdd5f/pkg/sql/logictest/testdata/logic_test/strict_ddl_atomicity#L13)):*
> ```
> 
> statement ok
> SET autocommit_before_ddl=off;
> ```

super nit: it might make sense to define this as one of the first statements in the test file, with a comment saying why.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141145)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @asg0451, @jeffswenson, and @spilchen)

___
*[`pkg/sql/logictest/testdata/logic_test/txn` line 1489 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTsIh82pa9EkPhzEK5:-OJ_03KZ-hyzYcq42Gvi:b-2mui3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/txn#L1489)):*
> The `CREATE SEQUENCE` statement commits, but the new transaction starts as read-only due to `default_transaction_read_only`. Should we ensure that when auto-committing, the new transaction inherits the read/write state of the one just committed?

i think we should avoid implementing that. it would get a little bit confusing to keep track of which per-transaction settings were inherited from somewhere else versus actually being configured by the user.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: Ensure we auto commit for first DDL in txn

Previously, when the autocommit_before_ddl session setting was enabled,
the first DDL statement in a multi-statement transaction did not trigger
an auto-commit, leading to the use of the legacy schema changer. This
behavior was observed in transactions such as:

```
SET autocommit_before_ddl = ON;
BEGIN;
DROP TABLE t1;
DROP TABLE t2;
```

This change will now ensure that we auto commit before the first drop.
This ensures that we'll use the DSC for both drops.

Epic: None
Release note: None

Closes #141034

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141145)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson and @rafiss)

___
*[`pkg/sql/logictest/testdata/logic_test/strict_ddl_atomicity` line 13 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJZziOX9QY5oeTpG7u1:-OJcUVxA8OQIM8Rlx8qf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef032150ea632680d52f7e29f8fe1fe4d5dbdd5f/pkg/sql/logictest/testdata/logic_test/strict_ddl_atomicity#L13)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: it might make sense to define this as one of the first statements in the test file, with a comment saying why.
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/txn` line 1489 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141145#-OJTsIh82pa9EkPhzEK5:-OJcW7h65mwn9Bkrw1V_:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/f94d1c795d7944b23d4491c56e468876741078e4/pkg/sql/logictest/testdata/logic_test/txn#L1489)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

> The `CREATE SEQUENCE` statement commits, but the new transaction starts as read-only due to `default_transaction_read_only`. Should we ensure that when auto-committing, the new transaction inherits the read/write state of the one just committed?

i think we should avoid implementing that. it would get a little bit confusing to keep track of which per-transaction settings were inherited from somewhere else versus actually being configured by the user.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

rafiss commented:
I added backport labels to this, as I think we'd want this behavior in current versions as well.

spilchen commented:
> I added backport labels to this, as I think we'd want this behavior in current versions as well.

Okay. I may hold off doing the backport for a few days to wait until the dust settles with this one. I don't anticipate anything major, but just in case.

spilchen commented:
TFTR!

bors r+
